### PR TITLE
[Layout][Heading] add ID prop to Heading, Layout.AnnotatedSection

### PR DESCRIPTION
**Pull request by brendanrygus**
### WHY are these changes introduced?
* In the SFN application, we want to link users from the home page to a section of the Settings page
* The `AnnotatedSection` subcomponents allow us to break up content of a Settings page, but do not support hash links
* The implicit layout used by `Layout` assumes that `Section` or `AnnotatedSection` will always be the direct child node of the parent, which does not allow consumers to wrap their content section in a labelled div

Example of broken layout with a wrapper div around the shipping section:
Screen Shot 2021-07-12 at 4 16 35 PM

### WHAT is this pull request doing?
* Adds an optional ID prop to the `Heading` component
* Adds an optional ID prop to the `Layout.AnnotatedSection` component, passed through to the primary Heading

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
**I couldn't get this working locally 🤔, but the changes seem pretty straightforward**
- [ ] ~For visual design changes, ping @ sarahill to update the Polaris UI kit~

### CODE REVIEW

**Comment by brendanrygus** [comment]
> Within the annotated section there's a `<Heading />` component with the section title. Would it make more sense to add the id there? I think that would be better for screen readers

Thanks @kyledurand, I had wondered about that as well. I was focusing more on the scroll position on my first pass. I'll push a quick update

File: src/components/Heading/Heading.tsx [hunk 1]
```diff
@@ -12,8 +12,14 @@ export interface HeadingProps {
   element?: HeadingTagName;
   /** The content to display inside the heading */
   children?: React.ReactNode;
+  /** A unique identifier for the heading, used for reference in anchor links  */
+  id?: string;
```
Comment (brendanrygus) [inline src/components/Heading/Heading.tsx:16]: @kyledurand Turns out the Heading didn't take an ID prop, so I updated the component and type docs.
If this looks good, let me know if there are any examples we should add to the README / code sample docs

File: UNRELEASED.md [hunk 2]
```diff
@@ -6,6 +6,9 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 
 ### Enhancements
 
+- [Layout] Add `id` prop for hash linking to `AnnotatedSection` components
+- [Heading] Add `id` prop
```
Comment (kyledurand) [inline UNRELEASED.md:10]: Can we change this to something like:
```suggestion
- Added `id` prop to `Layout` and `Heading` for hash linking ([#4307](https://github.com/Shopify/polaris-react/pull/4307))
```

To match our [guidelines](https://github.com/Shopify/polaris-react/blob/main/documentation/Versioning%20and%20changelog.md#user-content-component-name-formats)

**Review by kyledurand — COMMENTED** [review]
This is looking good. Can you add a quick test to heading as well? Just to make sure we don't remove the id in the future and cause a regression.

**Comment by brendanrygus** [comment]
Thanks for the speedy reviews @kyledurand!

This is my first PR here - I noticed that the contributing guide says: 
> “We’ll review your pull request and either merge it, request changes to it, or close it with an explanation.”

Let me know if you need anything else from my end - squashing commits, second approvals etc. - before it can be merged

**Pull request merged**